spirv_msl.cpp

```diff
@@ -9626,6 +9656,75 @@ void CompilerMSL::emit_instruction(const Instruction &instruction)
 		break;
 	}
 
+	case OpSDot:
+	case OpUDot:
+	case OpSUDot:
+	{
+		uint32_t result_type = ops[0];
+		uint32_t id = ops[1];
+		uint32_t vec1 = ops[2];
+		uint32_t vec2 = ops[3];
+
+		auto& input_type1 = expression_type(vec1);
+		auto& input_type2 = expression_type(vec2);
+
+		string vec1input, vec2input;
+		if ((input_type1.basetype == SPIRType::BaseType::UInt || input_type1.basetype == SPIRType::BaseType::Int) && input_type1.vecsize == 1)
+		{
+			string type = opcode == OpSDot || opcode == OpSUDot ? "char4" : "uchar4";
+			vec1input = join("as_type<", type, ">(", to_expression(vec1), ")");
+		} else
+			vec1input = to_expression(vec1);
+
+		if ((input_type2.basetype == SPIRType::BaseType::UInt || input_type2.basetype == SPIRType::BaseType::Int) && input_type2.vecsize == 1)
+		{
+			string type = opcode == OpSDot ? "char4" : "uchar4";
+			vec2input = join("as_type<", type, ">(", to_expression(vec2), ")");
+		} else
+			vec2input = to_expression(vec2);
+
+		auto& type = get<SPIRType>(result_type);
+		auto input_width = input_type1.vecsize == 1 ? 4 : input_type1.vecsize;
+		string exp = join("reduce_add(", type_to_glsl(type), std::to_string(input_width), "(", vec1input, ") * ", type_to_glsl(type), std::to_string(input_width), "(", vec2input, "))");
+		emit_op(result_type, id, exp, should_forward(vec1) && should_forward(vec2));
+		break;
+	}
+
+	case OpSDotAccSat:
+	case OpUDotAccSat:
+	case OpSUDotAccSat:
+	{
+		uint32_t result_type = ops[0];
+		uint32_t id = ops[1];
+		uint32_t vec1 = ops[2];
+		uint32_t vec2 = ops[3];
+		uint32_t acc = ops[4];
+
+		auto& input_type1 = expression_type(vec1);
+		auto& input_type2 = expression_type(vec2);
+
+		string vec1input, vec2input;
+		if ((input_type1.basetype == SPIRType::BaseType::UInt || input_type1.basetype == SPIRType::BaseType::Int) && input_type1.vecsize == 1)
+		{
+			string type = opcode == OpSDotAccSat || opcode == OpSUDotAccSat ? "char4" : "uchar4";
+			vec1input = join("as_type<", type, ">(", to_expression(vec1), ")");
+		} else
+			vec1input = to_expression(vec1);
+
+		if ((input_type2.basetype == SPIRType::BaseType::UInt || input_type2.basetype == SPIRType::BaseType::Int) && input_type2.vecsize == 1)
+		{
+			string type = opcode == OpSDotAccSat ? "char4" : "uchar4";
+			vec2input = join("as_type<", type, ">(", to_expression(vec2), ")");
+		} else
+			vec2input = to_expression(vec2);
+
+		auto& type = get<SPIRType>(result_type);
+		auto input_width = input_type1.vecsize == 1 ? 4 : input_type1.vecsize;
+		string exp = join("addsat(reduce_add(", type_to_glsl(type), std::to_string(input_width), "(", vec1input, ") * ", type_to_glsl(type), std::to_string(input_width), "(", vec2input, ")), ", to_expression(acc), ")");
```

Comment:
The result type cannot be used directly I think. If the opcode is SDotAccSat, you have to force the addition to happen in int, not uint. SPIR-V can mix and match like that. UDotAccSat seems to have language forcing signedness of 0, so that's probably fine as-is ...

Reply:
Yes, I think you're right. I guess I'll just have to add some logic to generate "uint4" when `type.width == 32` and `input_type1.vecsize == 4` and its `OpUDot`. When it's `OpSDot` it'll always be `int4` regardless of the return type. Do I need to take anything else into account? Though interesting that the Vulkan CTS tests don't test if that is handled correctly...

Reply:
Oh wait you're talking about the saturated addition? The SPIR-V spec says `The type of Accumulator must be the same as Result Type.`. So I'm pretty sure it can be used.

I thought you were talking about the result type of the component-wise multiplication. I'm not entirely sure, but I thought if its OpSDot and the result type is uint, the multiplication and reduce_add would still have to be done on signed integers, no? So me casting it always to the vector version of the result type could possibly be wrong because I'm possibly not sign-extending, right?

Reply:
Using signedness in OpTypeInt is somewhat optional in SPIR-V and only really applies in specific circumstances.

Reply:
That would require a `as_type<int4>(x)` then, no?

Reply:
Nah, just a regular functional cast will do.

Reply:
I'm still not sure I can wrap my head around the entirety of this issue tbh. However, I've changed the code to always cast to the result type but in the case of OpSDot and the first vector of OpSUDot it'll cast to the signed version thereof. I think that solves the issue?